File: leetcode/python/303. 区域和检索 - 数组不可变.py
#!/bin/bash python
"""
303. 区域和检索 - 数组不可变

思路：
一维前缀和
"""

class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        一维dp，前缀和
        """
        self.dp = nums
        for i in range(1, len(nums)):
            self.dp[i] += self.dp[i - 1]
            
    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        if self.dp == [] or i > j or i < 0:
            return 0
        elif i == 0:
            return self.dp[j]
        else:
            return self.dp[j] - self.dp[i - 1]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)

File: leetcode/cpp/303. 区域和检索 - 数组不可变.cpp
class NumArray {
private:
    vector<int> dp;
public:
    // 前缀和
    NumArray(vector<int>& nums) {
        int n = nums.size();
        dp.resize(n + 1, 0);
        for (int i = 0; i < n; i++) {
            dp[i + 1] = dp[i] + nums[i];
        }
    }
    
    int sumRange(int i, int j) {
        return dp[j + 1] - dp[i];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(i,j);
 */